File: app/src/main/java/com/coinninja/coinkeeper/service/runner/FulfillSentInvitesRunner.java
package com.coinninja.coinkeeper.service.runner;


import android.content.Context;
import android.content.res.Resources;

import com.coinninja.coinkeeper.di.interfaces.ApplicationContext;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.helpers.TransactionHelper;

import java.util.List;

import javax.inject.Inject;

public class FulfillSentInvitesRunner implements Runnable {
    private final Resources resources;

    private final TransactionHelper transactionHelper;
    private final SentInvitesStatusGetter sentInvitesStatusGetter;
    private final SentInvitesStatusSender sentInvitesStatusSender;
    private final BroadcastBtcInviteRunner broadcastBtcInviteRunner;

    @Inject
    FulfillSentInvitesRunner(@ApplicationContext Context context,
                             TransactionHelper transactionHelper,
                             SentInvitesStatusGetter sentInvitesStatusGetter,
                             SentInvitesStatusSender sentInvitesStatusSender,
                             BroadcastBtcInviteRunner broadcastBtcInviteRunner) {

        resources = context.getResources();
        this.broadcastBtcInviteRunner = broadcastBtcInviteRunner;
        this.sentInvitesStatusGetter = sentInvitesStatusGetter;
        this.transactionHelper = transactionHelper;
        this.sentInvitesStatusSender = sentInvitesStatusSender;
    }

    @Override
    public void run() {

        //Step 1. grab all sent invites from server, save/update the ones that now have an address
        sentInvitesStatusGetter.run();

        //Step 2. get any sent invites that do not have a tx id but have an address
        List<InviteTransactionSummary> unfulfilledTransactions = transactionHelper.gatherUnfulfilledInviteTrans();
        broadcastRealTxForInvites(unfulfilledTransactions);

        //Step 3. report to coinninja server of any invites that have been newly fulfilled (with TX ID)
        sentInvitesStatusSender.run();
    }

    public void broadcastRealTxForInvites(List<InviteTransactionSummary> unfulfilledTransactions) {
        for (InviteTransactionSummary transaction : unfulfilledTransactions) {
            if (alreadyHasTxId(transaction))
                continue;//if this invite already has a tx id then skip it

            broadcastBtcInviteRunner.setInvite(transaction);
            broadcastBtcInviteRunner.run();

            sleep(1000);//see if this fixes spending change address issue?
        }
    }

    protected boolean alreadyHasTxId(InviteTransactionSummary transaction) {
        String currentTXID = transaction.getBtcTransactionId();
        return !(currentTXID == null || currentTXID.isEmpty());
    }

    private void sleep(long ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/service/runner/FulfillSentInvitesRunner.kt
package com.coinninja.coinkeeper.service.runner


import app.coinninja.cn.thunderdome.repository.ThunderDomeRepository
import app.dropbit.annotations.Mockable
import app.dropbit.commons.util.isNotNullOrEmpty
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary
import com.coinninja.coinkeeper.model.db.enums.BTCState
import com.coinninja.coinkeeper.model.helpers.InviteTransactionSummaryHelper
import com.coinninja.coinkeeper.service.client.SignedCoinKeeperApiClient
import javax.inject.Inject

@Mockable
class FulfillSentInvitesRunner @Inject internal constructor(
        internal val inviteTransactionSummaryHelper: InviteTransactionSummaryHelper,
        internal val sentInvitesStatusGetter: SentInvitesStatusGetter,
        internal val sentInvitesStatusSender: SentInvitesStatusSender,
        internal val broadcastBtcInviteRunner: BroadcastBtcInviteRunner,
        internal val thunderDomeRepository: ThunderDomeRepository,
        internal val cnClient: SignedCoinKeeperApiClient
) : Runnable {

    override fun run() {

        //Step 1. grab all sent invites from server, save/update the ones that now have an address
        sentInvitesStatusGetter.run()

        //Step 2. addressForPubKey any sent invites that do not have a tx id but have an address
        broadcastInvitesOnBlockchain(inviteTransactionSummaryHelper.unfulfilledSentInvites)
        broadcastInvitesOnThunderDome(inviteTransactionSummaryHelper.unfulfilledLightningSentInvites)

        //Step 3. report to coinninja server of any invites that have been newly fulfilled (with TX ID)
        sentInvitesStatusSender.run()
    }

    internal fun broadcastInvitesOnThunderDome(invites: List<InviteTransactionSummary>) {
        invites.forEach { invite ->
            if (invite.address.isNotNullOrEmpty()) {
                thunderDomeRepository.pay(invite.address, invite.valueSatoshis)?.let { invoice ->
                    invoice.id.let { id ->
                        invite.btcState = BTCState.FULFILLED
                        invite.update()
                        cnClient.updateInviteStatusCompleted(invite.serverId, id.split(":")[0])
                    }

                }
            }
        }
    }

    private fun broadcastInvitesOnBlockchain(unfulfilledTransactions: List<InviteTransactionSummary>) {
        for (transaction in unfulfilledTransactions) {
            if (alreadyHasTxId(transaction)) continue

            broadcastBtcInviteRunner.invite = transaction
            broadcastBtcInviteRunner.run()
        }
    }

    fun alreadyHasTxId(transaction: InviteTransactionSummary): Boolean {
        val currentTXID = transaction.btcTransactionId
        return !(currentTXID == null || currentTXID.isEmpty())
    }
}
